### PR TITLE
Add deterministic, pure-Python roi_align implementation

### PR DESCRIPTION
**Pull request by ezyang**
See
https://dev-discuss.pytorch.org/t/a-pure-python-implementation-of-roi-align-that-looks-just-like-its-cuda-kernel/1266
for discussion and motivation.

Signed-off-by: Edward Z. Yang <ezyang@meta.com>

### CODE REVIEW

File: torchvision/ops/roi_align.py [hunk 1]
```diff
@@ -54,12 +187,14 @@ def roi_align(
...
         _log_api_usage_once(roi_align)
-    _assert_has_ops()
     check_roi_boxes_shape(boxes)
     rois = boxes
     output_size = _pair(output_size)
     if not isinstance(rois, torch.Tensor):
         rois = convert_boxes_to_roi_format(rois)
+    if not torch.jit.is_scripting():
+        if not _has_ops() or (torch.are_deterministic_algorithms_enabled() and input.is_cuda):
+            return _roi_align(input, rois, spatial_scale, output_size[0], output_size[1], sampling_ratio, aligned)
```
Comment (NicolasHug) [inline torchvision/ops/roi_align.py:197]: call to `_assert_has_ops()` was removed above so we should probably put it back after those `if` blocks

```suggestion
            return _roi_align(input, rois, spatial_scale, output_size[0], output_size[1], sampling_ratio, aligned)
    _assert_has_ops()
```

File: torchvision/ops/roi_align.py [hunk 2]
```diff
@@ -1,16 +1,149 @@
...
 from torch.jit.annotations import BroadcastingList2
 from torch.nn.modules.utils import _pair
-from torchvision.extension import _assert_has_ops
+from torchvision.extension import _has_ops
 
 from ..utils import _log_api_usage_once
 from ._utils import check_roi_boxes_shape, convert_boxes_to_roi_format
 
 
+# NB: all tensor inputs
```
Comment (NicolasHug) [inline torchvision/ops/roi_align.py:15]: Not sure what you mean by this comment? Maybe this?


```suggestion
# NB: all tensor are inputs
```

File: torchvision/ops/roi_align.py [hunk 3]
```diff
@@ -1,16 +1,149 @@
...
+    if torch.is_autocast_enabled() and tensor.is_cuda and tensor.dtype != torch.double:
+        return tensor.float()
+    else:
+        return tensor
+
+
+# This is a slow but pure Python and differentiable implementation of
+# roi_align.  It potentially is a good basis for Inductor compilation
+# (but I have not benchmarked it) but today it is solely used for the
+# fact that its backwards can be implemented deterministically.
```
Comment (NicolasHug) [inline torchvision/ops/roi_align.py:70]: Nit, for future ref

```suggestion
# fact that its backwards can be implemented deterministically,
# which is needed for the PT2 benchmark suite.
```

File: test/test_ops.py [hunk 4]
```diff
@@ -140,7 +157,8 @@ def test_torch_fx_trace(self, device, x_dtype=torch.float, rois_dtype=torch.floa
     @pytest.mark.parametrize("seed", range(10))
     @pytest.mark.parametrize("device", cpu_and_gpu())
     @pytest.mark.parametrize("contiguous", (True, False))
-    def test_backward(self, seed, device, contiguous):
+    @pytest.mark.parametrize("deterministic", (False,))
+    def test_backward(self, seed, device, contiguous, deterministic):
```
Comment (NicolasHug) [inline test/test_ops.py:161]: I think we can get rid of the parametrization here, and do as you did above in `test_forward()`, i.e. just add the param with a default

```suggestion
    def test_backward(self, seed, device, contiguous, deterministic=False):
```

File: test/test_ops.py [hunk 5]
```diff
@@ -402,21 +421,42 @@ def test_boxes_shape(self):
...
+                aligned=aligned,
+                x_dtype=x_dtype,
+                rois_dtype=rois_dtype,
             )
 
+    @pytest.mark.parametrize("seed", range(10))
+    @pytest.mark.parametrize("device", cpu_and_gpu())
+    @pytest.mark.parametrize("contiguous", (True, False))
+    @pytest.mark.parametrize("deterministic", (True, False))
+    def test_backward(self, seed, device, contiguous, deterministic):
```
Comment (NicolasHug) [inline test/test_ops.py:457]: NIT: `test_backward()` isn't as slow as `test_forward()`, but we may still want to add the same `pytest.skip` here?

File: torchvision/ops/roi_align.py [hunk 6]
```diff
@@ -1,16 +1,149 @@
...
+#
+# It is transcribed directly off of the roi_align CUDA kernel, see
+# https://dev-discuss.pytorch.org/t/a-pure-python-implementation-of-roi-align-that-looks-just-like-its-cuda-kernel/1266
+@torch._dynamo.allow_in_graph
+def _roi_align(input, rois, spatial_scale, pooled_height, pooled_width, sampling_ratio, aligned):
+    from functorch.dim import dims
```
Comment (NicolasHug) [inline torchvision/ops/roi_align.py:76]: Any reason to lazy import? Isn't the `functorch` namespace always available now?

**Review by NicolasHug — APPROVED** [review]
Thanks @ezyang . PR Looks good, I just have minor Qs and comments below. There's nothing critical so I'll approve now.

I haven't checked the correctness of the new implementation too closely, I'm mostly relying on our tests for that. Out of precaution I parametrized the `test_forward()` over 50 seeds locally and they all passed.

File: torchvision/ops/roi_align.py (hunk 6 above)
Comment (zou3519) [reply torchvision/ops/roi_align.py:77]: A better reason is that `import functorch.dim` monkey-patches torch and we really do not want to monkey patch torch by default.

Comment (ezyang) [reply torchvision/ops/roi_align.py]: For safety purposes, I hand-inserted all of the expands necessary to remove the first-class dims impl. We should still keep the first class dims version around for documentary purposes though.
